plugins/woocommerce/client/blocks/assets/js/atomic/blocks/product-elements/image/block.tsx

```diff
@@ -101,6 +102,28 @@ const Image = ( {
 type Props = BlockAttributes &
 	HTMLAttributes< HTMLDivElement > & { style?: Record< string, unknown > };
 
+// props.product is not listed in the BlockAttributes explicitly,
+// but it is implicitly passed from the All Products block.
+// This is what distinguishes this block from the other usage of the Product Image component.
+const displayLegacySaleBadge = ( props: Props ) => {
+	const { product } = props;
+	const isInAllProducts = ! isEmpty( product );
+
+	if ( isInAllProducts ) {
+		// If the All Products block is pristine, it doesn't have a showSaleBadge attribute
+		// but it was `true` by default.
+		const trueByDefault = props.showSaleBadge === undefined;
+
+		// If the All Products block was edited, it will have a showSaleBadge attribute
+		// that we should respect.
+		const trueAfterEdit = props.showSaleBadge === true;
+
+		return trueByDefault || trueAfterEdit;
```

Comment:
Nitpick and probably a matter of style preference, but you could return earlier if it was `trueByDefault`. As a reader, I'm just _ever-so-slightly_ confused by why we check if it's `trueAfterEdit` and I needed that extra second to understand it was just for style preference.

Not a big deal, but think about it.

Reply:
Actually I refactored it to

```
	// If the block is pristine, it doesn't have a showSaleBadge attribute
	// but it is supposed to be`true` by default.
	if ( showSaleBadge === undefined ) {
		return true;
	}

	// If the block was edited, it will have a showSaleBadge attribute
	// that we should respect.
	return showSaleBadge;
```

I guess it's a nice middle ground 😄

Reply:
Done https://github.com/woocommerce/woocommerce/pull/57889/commits/d172eef438ac47d24d7a6b5a15e736b340241f52